README.md
# BNPP IRB for Magento

BNPP IRB for Magento is an open source plugin that links e-commerce websites based on Magento to BNPP IRB secured payment gateway developped by [Lyra Network](https://www.lyra-network.com/).

Namely, it enables the following payment methods :
* BNPP IRB - Standard credit card payment
* BNPP IRB - Credit card payment in installments

## Installation & upgrade

- Create app/code/Lyranetwork/Bnppirb folder if not exists.
- Unzip module in your Magento 2 app/code/Lyranetwork/Bnppirb folder.
- Open command line and change to Magento installation root directory.
- Enable module: php bin/magento module:enable --clear-static-content Lyranetwork_Bnppirb
- Upgrade database: php bin/magento setup:upgrade
- Re-run compile command: php bin/magento setup:di:compile
- Update static files by: php bin/magento setup:static-content:deploy [locale]

In order to deactivate the module: php bin/magento module:disable --clear-static-content Lyranetwork_Bnppirb

## Configuration

- In Magento 2 administration interface, browse to "STORES > Configuration" menu
- Click on "Payment Methods" link under "SALES" section
- Expand BNPP IRB payment method to enter your gateway credentials
- Refresh invalidated Magento cache afeter config saved. 

## License

Each BNPP IRB payment module source file included in this distribution is licensed under Open Software License (OSL 3.0).

Please see LICENSE.txt for the full text of the OSL 3.0 license. It is also available through the world-wide-web at this URL: https://opensource.org/licenses/osl-3.0.php.


CHANGELOG.md
2.3.1, 2018-07-06:
- [shatwo]Enable SHA-256 signature algorithm by default.
- Ignore spaces at the beginning and the end of certificates on return signature processing.

2.3.0, 2018-05-23:
- Enable signature algorithm selection (SHA-1 or SHA-256).
- Improve backend configuration screen.

2.2.0, 2018-03-19:
- Display card brand user choice if any in backend order details.

2.1.4, 2018-03-08:
- Bug fix: Check value type in logo uploader to avoid a PHP warning during import data from configuration files.
- Bug fix: Added the component load order in etc/module.xml to avoid an exception during Magento installation using comand-line interface.
- [technical]Manage enabled/disabled features by plugin variant.

2.1.3, 2017-10-16:
- Compatibility with Magento 2.2 version (method signature, namespace use, JSON unserialize and layout init).

2.1.2, 2017-08-14:
- Bug fix: compilation problem relative to class constructor parameters.
- Possibility to view payment page within pop-in for standard payment (iframe mode).
- Possibility to select payment card type on merchant website (for both standard payment and payment in installments).
- Save both converted and original paid amounts in Magento transaction details.
- Use of protected variables (instead of private) to facilitate module code extension.
- [multi]Register details about all payment installments.

2.1.1, 2017-01-13:
- Bug fix: correction of an error when returning to store using browser backward button.
- [multi]Bug fix: selected payment in installments option is not considered since Magento 2.1.3.
- Update module structure and code to fulfill Magento marketplace requirements.
- Using "lyranetwork" as vendor name instead of "lyra" (already used in Magento Marketplace).

2.1.0, 2016-11-05:
- Bug fix: error relative to CMS version detection since Magento V 2.1.
- Bug fix: notify URL is not displayed correctly in module backend since Magento V 2.1.
- [multi]Bug fix: multiple payment sub-module dos not work since Magento V 2.1 (selected payment option is not considered in frontend).
- Remove control over certificate format modified on the gateway.
- Possibility to refund payments from Magento backend via WS.
- Backend payments in MOTO mode.
- Upgrade supported card types list.

2.0.1, 2016-06-02:
- Bug fix: correction of an error in backend order creation page when module is enabled.
- Bug fix: use of _scope parameter (instead of _store) in store URLs to redirect to the correct store (in multistore mode).
- Bug fix: error occures when saving module settings without resfreshing cache (if asked by Magento).
- Bug fix: capture delay in sub-modules not considered if equals 0.
- [multi]Bug fix: payment in installments did not work (always processed as single).
- Dispatch event order_cancel_after after order cancellation.
- Check if current quote currency is supported before to check base currency.
- Improve logging system and log request format errors.
- Update german language file.
- Improve of label fields display on admin panel.
- [multi]Do not delete virtual multi payment methods (bnppirb_multi_Nx) to avoid errors when viewing orders paid with these methods.

2.0.0, 2016-03-10:
- New BNPP IRB payment module for magento 2.


Helper/Data.php
<?php
/**
 * BNPP IRB V2-Payment Module version 2.3.1 for Magento 2.x. Support contact : csp.monetique.afrique@bnpparibas.com.
 *
 * NOTICE OF LICENSE
 *
 * This source file is licensed under the Open Software License version 3.0
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/osl-3.0.php
 *
 * @author    Lyra Network (http://www.lyra-network.com/)
 * @copyright 2014-2018 Lyra Network and contributors
 * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 * @category  payment
 * @package   bnppirb
 */
namespace Lyranetwork\Bnppirb\Helper;

use Magento\Store\Model\ScopeInterface;
use Magento\Framework\App\Config\ScopeConfigInterface;
use Magento\Framework\App\Filesystem\DirectoryList;

class Data extends \Magento\Framework\App\Helper\AbstractHelper
{

    const METHOD_STANDARD = 'bnppirb_standard';

    const METHOD_MULTI = 'bnppirb_multi';

    const METHOD_CHOOZEO = 'bnppirb_choozeo';

    const METHOD_GIFT = 'bnppirb_gift';

    const METHOD_ONEY = 'bnppirb_oney';

    const METHOD_PAYPAL = 'bnppirb_paypal';

    const METHOD_SOFORT = 'bnppirb_sofort';

    const METHOD_POSTFINANCE = 'bnppirb_postfinance';

    /**
     *
     * @var array a global var to easily enable/disable features
     */
    public static $pluginFeatures = [
        'qualif' => false,
        'prodfaq' => true,
        'restrictmulti' => false,
        'shatwo' => true,

        'multi' => true,
        'choozeo' => false
    ];

    /**
     *
     * @var \Magento\Framework\ObjectManagerInterface
     */
    protected $objectManager;

    /**
     *
     * @var \Magento\Store\Model\StoreManagerInterface
     */
    protected $storeManager;

    /**
     *
     * @var \Magento\Framework\App\MaintenanceMode
     */
    protected $maintenanceMode;

    /**
     *
     * @var \Magento\Config\Model\ResourceModel\Config
     */
    protected $resourceConfig;

    /**
     *
     * @var \Magento\Framework\Filesystem
     */
    protected $filesystem;

    /**
     *
     * @var \Magento\Config\Model\Config\Structure
     */
    protected $configStructure;

    /**
     *
     * @var \Lyranetwork\Bnppirb\Model\Logger\Bnppirb
     */
    protected $logger;

    /**
     *
     * @var \Magento\Framework\App\State
     */
    protected $appState;

    /**
     *
     * @var \Zend\Http\PhpEnvironment\RemoteAddress
     */
    protected $remoteAddress;

    /**
     *
     * @var \Magento\Framework\Filesystem\Io\File
     */
    protected $file;

    /**
     *
     * @param \Lyranetwork\Bnppirb\Helper\Context $context
     * @param \Magento\Framework\ObjectManagerInterface $objectManager
     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
     * @param \Magento\Framework\App\MaintenanceMode $maintenanceMode
     * @param \Magento\Config\Model\ResourceModel\Config $resourceConfig
     * @param \Magento\Framework\Filesystem $filesystem
     * @param \Magento\Config\Model\Config\Structure $configStructure
     * @param \Lyranetwork\Bnppirb\Model\Logger\Bnppirb
     * @param \Magento\Framework\App\State $appState
     * @param \Zend\Http\PhpEnvironment\RemoteAddress $remoteAddress
     * @param \Magento\Framework\Filesystem\Io\File $file
     */
    public function __construct(
        \Lyranetwork\Bnppirb\Helper\Context $context,
        \Magento\Framework\ObjectManagerInterface $objectManager,
        \Magento\Store\Model\StoreManagerInterface $storeManager,
        \Magento\Framework\App\MaintenanceMode $maintenanceMode,
        \Magento\Config\Model\ResourceModel\Config $resourceConfig,
        \Magento\Framework\Filesystem $filesystem,
        \Magento\Config\Model\Config\Structure $configStructure,
        \Lyranetwork\Bnppirb\Model\Logger\Bnppirb $logger,
        \Magento\Framework\App\State $appState,
        \Zend\Http\PhpEnvironment\RemoteAddress $remoteAddress,
        \Magento\Framework\Filesystem\Io\File $file
    ) {
        parent::__construct($context);

        $this->objectManager = $objectManager;
        $this->storeManager = $storeManager;
        $this->maintenanceMode = $maintenanceMode;
        $this->resourceConfig = $resourceConfig;
        $this->filesystem = $filesystem;
        $this->configStructure = $configStructure;
        $this->logger = $logger;
        $this->appState = $appState;
        $this->remoteAddress = $remoteAddress;
        $this->file = $file;
    }

    /**
     * Shortcut method to get general BNPP IRB module configuration.
     *
     * @param string $field
     * @param int|string|null|\Magento\Store\Model\Store $storeId
     * @return mixed
     */
    public function getCommonConfigData($field, $storeId = null)
    {
        if ($storeId === null && $this->isBackend()) {
            $storeId = $this->getCheckoutStoreId();
        }

        return $this->scopeConfig->getValue('bnppirb/general/' . $field, ScopeInterface::SCOPE_STORE, $storeId);
    }

    /**
     * Return user's IP Address.
     *
     * @return string
     */
    public function getIpAddress()
    {
        return $this->remoteAddress->getIpAddress();
    }

    /**
     * Get the complete payment return URL.
     *
     * @param int $storeId the ID of the store
     * @return string
     */
    public function getReturnUrl($storeId = null)
    {
        $params = [];
        $params['_nosid'] = true;
        $params['_secure'] = $this->getStore($storeId)->isCurrentlySecure();

        if ($storeId) {
            $params['_scope'] = $storeId;
        }

        return $this->_getUrl('bnppirb/payment/response', $params);
    }

    /**
     * Return true if this is a backend session.
     *
     * @return bool
     */
    public function isBackend()
    {
        return $this->appState->getAreaCode() == \Magento\Backend\App\Area\FrontNameResolver::AREA_CODE;
    }

    /**
     * Return true if SSL is enabled for current store.
     *
     * @return bool
     */
    public function isCurrentlySecure()
    {
        return $this->storeManager->getStore()->isCurrentlySecure();
    }

    /**
     * Return checkout session.
     *
     * @return Magento\Backend\Model\Session|Magento\Backend\Model\Session\Quote
     */
    public function getCheckout()
    {
        $sessionClass = \Magento\Checkout\Model\Session::class;
        if ($this->isBackend()) {
            $sessionClass = \Magento\Backend\Model\Session\Quote::class;
        }

        return $this->objectManager->get($sessionClass);
    }

    /**
     * Return current checkout store.
     *
     * @return int
     */
    public function getCheckoutStoreId()
    {
        $session = $this->getCheckout();
        return $session->getStoreId();
    }

    /**
     * Return store obeject by ID.
     *
     * @param int $storeId
     * @return \Magento\Store\Model\Store
     */
    public function getStore($storeId)
    {
        return $this->storeManager->getStore($storeId);
    }

    /**
     * Return current checkout quote.
     *
     * @return \Magento\Quote\Model\Quote
     */
    public function getCheckoutQuote()
    {
        $session = $this->getCheckout();
        return $session->getQuote();
    }

    /**
     * Return true if Magento shop is in maintenance mode.
     *
     * @return bool
     */
    public function isMaintenanceMode()
    {
        return $this->maintenanceMode->isOn($this->getIpAddress());
    }

    /**
     * Return true file exists.
     *
     * @param string $file
     * @param bool $onlyFile
     * @return bool
     */
    public function fileExists($file, $onlyFile = true)
    {
        return $this->file->fileExists($file, $onlyFile);
    }

    /**
     * Check if image file is uploaded to media directory.
     *
     * @param string $fileName
     * @return string
     */
    public function isUploadFileImageExists($fileName)
    {
        $filePath = $this->filesystem->getDirectoryRead(DirectoryList::MEDIA)->getAbsolutePath('bnppirb/images/' . $fileName);
        return $this->fileExists($filePath);
    }

    /**
     * Check if image file is published to pub/static directory.
     *
     * @param string $fileName
     * @return string
     */
    public function isPublishFileImageExists($fileName)
    {
        $filePath = $this->filesystem->getDirectoryRead(DirectoryList::STATIC_VIEW)->getAbsolutePath($fileName);
        return $this->fileExists($filePath);
    }

    /**
     * Returns a configuration parameter from XML files.
     *
     * @param string $path
     * @return string
     */
    public function getGroupTitle($path)
    {
        // $path is as payment_[lang]/bnppirb/bnppirb_[group]/bnppirb_[sub_group]
        $parts = explode('/', $path);
        $parentPath = 'payment/' . $parts[1] . '/' . $parts[2]; // we need the second level group

        return __($this->configStructure->getElement($parentPath)->getLabel())->render();
    }

    /**
     * Add a model config parameter for each of given $options (multi payment options).
     *
     * @param array[string][mixed] $options
     */
    public function updateMultiPaymentModelConfig($options)
    {
        foreach ($options as $option) {
            $this->resourceConfig->saveConfig(
                'payment/bnppirb_multi_' . $option['count'] . 'x/model',
                \Lyranetwork\Bnppirb\Model\Method\Multix::class,
                ScopeConfigInterface::SCOPE_TYPE_DEFAULT,
                0
            );
        }
    }

    /**
     * Get multi payment method models.
     *
     * @return array[int] $options
     */
    public function getMultiPaymentModelConfig()
    {
        // retrieve DB connection
        $connection = $this->resourceConfig->getConnection();

        $select = $connection->select()
            ->from($this->resourceConfig->getMainTable())
            ->where('path LIKE ?', 'payment/bnppirb\_multi\_%x/model');

        return $connection->fetchAll($select);
    }

    /**
     * Check if server has requirements to do WS operations.
     *
     * @throws \Lyranetwork\Bnppirb\Model\WsException
     */
    public function checkWsRequirements()
    {
        if (! extension_loaded('soap')) {
            throw new \Lyranetwork\Bnppirb\Model\WsException(
                'SOAP extension for PHP must be enabled on the server in order to use BNPP IRB web services.'
            );
        }

        if (! extension_loaded('openssl')) {
            throw new \Lyranetwork\Bnppirb\Model\WsException(
                'OPENSSL extension for PHP must be enabled on the server in order to use BNPP IRB web services.'
            );
        }
    }

    /**
     * Unserialize data using JSON or PHP unserialize function if error.
     *
     * @param string $string
     * @return mixed
     */
    public function unserialize($string)
    {
        $result = json_decode($string, true);

        if (json_last_error() !== JSON_ERROR_NONE) {
            // Magento 2.1.x : try PHP serialization
            $result = @unserialize($string);
        }

        return $result;
    }

    /**
     * Log function.
     *
     * @param string $message
     * @param string $level
     */
    public function log($message, $level = \Psr\Log\LogLevel::INFO)
    {
        if (! $this->getCommonConfigData('enable_logs')) {
            return;
        }

        $currentMethod = $this->getCallerMethod();

        $log = '';
        $log .= 'BNPP IRB 2.3.1';
        $log .= ' - ' . $currentMethod;
        $log .= ' : ' . $message;

        $this->logger->log($level, $log);
    }

    /**
     * Find the name of the method that called the log method.
     *
     * @return string|null
     */
    private function getCallerMethod()
    {
        $traces = debug_backtrace();

        if (isset($traces[2])) {
            return $traces[2]['class'] . '::' . $traces[2]['function'];
        }

        return null;
    }
}
